### PR TITLE
Volatile db

### PR DESCRIPTION
**Pull request by kderme**
This pr introduces volatile db and adds some tests

The db is a key-value store of binary blocks and is parametric on the key of blocks, named blockId. The only constraints are that one must provide
a function (toSlot :: blockId -> Slot), as well as an Ord instance of blockId.
The database uses in memory indexes, which are created on each reopening. reopening
includes parsing all blocks of the dbFolder, so it can be an expensive operation
if the database gets big. That's why the intention of this db is to be used for only
the tip of the blockchain, when there is still volatility on which blocks are included.
The db is agnostic to the format of the blocks, so a parser must be provided.
In addition to getBlock and putBlock, the db provides also the ability to garbage-collect
old blocks. The actual garbage-collection happens in terms of files and not blocks: a file
is deleted/garbage-collected only if its latest block is old enough. A block is old enough
if its toSlot value is old enough and not based on its Ord instance. This type of garbage
collection makes the deletion of blocks depend on the number of blocks we insert on
each file, as well as the order of insertion, so it's not deterministic on blocks themselves.

related issue: https://github.com/input-output-hk/ouroboros-network/issues/202

### CODE REVIEW

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs [hunk 1]
```diff
@@ -0,0 +1,469 @@
...
+import           Ouroboros.Storage.VolatileDB.API
+import           Ouroboros.Storage.VolatileDB.Util
+import           Ouroboros.Storage.Util.ErrorHandling (ErrorHandling (..))
+import qualified Ouroboros.Storage.Util.ErrorHandling as EH
+
+{------------------------------------------------------------------------------
+  Main Types
+------------------------------------------------------------------------------}
+
+-- This is only visible for testing.
```
Comment (kderme) [inline ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs]: this comment is no longer necessary

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs [hunk 2]
```diff
@@ -0,0 +1,469 @@
...
+data VolatileDBEnv m blockId = VolatileDBEnv {
+      _dbInternalState  :: !(TMVar m (Maybe (InternalState m blockId)))
+    , _dbFolder         :: !FsPath
+    , _maxBlocksPerFile :: !Int
+    , _parser           :: !(Parser m blockId)
+    , _toSlot           :: (blockId -> EpochSlot)
+    }
+
+-- This is only visible for testing.
```
Comment (kderme) [inline ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs]: this comment is no longer necessary

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs [hunk 3]
```diff
@@ -0,0 +1,469 @@
...
+            && _currentWriteOffset x == _currentWriteOffset y
+            && _currentNextId x == _currentNextId y
+            && _currentBlockId x == _currentBlockId y
+            && _currentMap x == _currentMap y
+            && _currentRevMap x == _currentRevMap y
+
+instance Show blockId => Show (InternalState m blockId) where
+    show InternalState{..} = show (_currentWritePath, _currentWriteOffset, _currentNextId, _currentBlockId, _currentMap, _currentRevMap)
+
+instance Eq blockId => Eq (VolatileDBError blockId) where
```
Comment (kderme) [inline ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs]: move this back to types

**Comment by edsko** [comment]
Reviewed in hangout.

**Comment by kderme** [comment]
@edsko changes since last review:
- added many comments
- some new tests especially related to reopening and corruptions or errors
- beautyfying
- some test improvements
- use Data.Map.withoutKeys for gc

TODO yet 
- remove dbFolder
- be more lenient on blocksPerFile requirements
I have added TODOs in code for the above. Could I handle them on separate follow-up prs? I'm still putting some thought on them.

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs [hunk 4]
```diff
@@ -0,0 +1,557 @@
...
+             -> BS.Builder
+             -> m ()
+putBlockImpl hasFS@HasFS{..} err toSlot env@VolatileDBEnv{..} blockId builder = do
+    modifyState hasFS err env $ \st@InternalState{..} -> do
+        case Map.lookup blockId _currentRevMap of
+            Just _ -> return (st, ()) -- trying to put an existing blockId is a no-op.
+            Nothing -> do
+                (mmaxSlot, nBlocks, fileMp) <-
+                    case Map.lookup _currentWritePath _currentMap of
+                        Nothing -> EH.throwError err $ UndisputableLookupError _currentWritePath _currentMap
```
Comment (edsko) [inline ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs]: Can add a TODO here also to make this an internal function (and use `error`).

Comment (kderme) [reply ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs]: @edsko You mean `error` instead of `EH.throwError`?

Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs]: We discussed this in the call no? This is an internal invariant violation, it's not an error that should ever happen? Am I confusing it with something else?

Comment (kderme) [reply ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs]: Yes that's true. I just did't remember that on internal invariant violations we should not follow the same approach as other errors. I actually have some tests with this error: I intentionally corrupt the internal state and check if this error is indeed raised. Maybe `error` would make these tests a bit weird to write?

Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs]: I think you can just delete those tests. Invariants are there for a reason; seeing how the code behaves when the invariants are broken isn't very useful, we can't say all that much meaningful about it.

If you are worried about the invariant being broken, a better test is to check for that instead: no matter what pattern of calls into the API we have, the internal invariant is always satisfied. So, write the invariant as a function, and check it. That _would_ be a very useful test.

**Pull request merged**